src/adapter/src/coord/introspection.rs

```diff
@@ -86,24 +87,34 @@ pub fn auto_run_on_introspection<'a, 's>(
 pub fn check_cluster_restrictions(
     catalog: &impl SessionCatalog,
     plan: &Plan,
-    depends_on: &Vec<GlobalId>,
 ) -> Result<(), AdapterError> {
     // We only impose restrictions if the current cluster is the introspection cluster.
     let cluster = catalog.active_cluster();
     if cluster != MZ_INTROSPECTION_CLUSTER.name {
         return Ok(());
     }
 
-    // Allows explain queries.
-    if let Plan::Explain(_) = plan {
-        return Ok(());
-    }
+    // Only continue, and check restrictions, if a Plan would run some computation on the cluster.
+    //
+    // Note: We get the dependencies from the Plans themselves, because it's only after planning
+    // that we actually know what objects we'll need to reference.
+    //
+    // Note: Creating other objects like Materialized Views is prevented elsewhere. We define the
+    // 'mz_introspection' cluster to be "read-only", which restricts these actions.
+    let depends_on: Box<dyn Iterator<Item = GlobalId>> = match plan {
+        Plan::ReadThenWrite(plan) => Box::new(plan.selection.depends_on().into_iter()),
+        Plan::Subscribe(plan) => match plan.from {
+            SubscribeFrom::Id(id) => Box::new(std::iter::once(id)),
+            SubscribeFrom::Query { ref expr, .. } => Box::new(expr.depends_on().into_iter()),
+        },
+        Plan::Peek(plan) => Box::new(plan.source.depends_on().into_iter()),
+        _ => return Ok(()),
+    };
```

Comment:
@ParkMyCar  I just realized I think this probably needs to include `Plan::Insert`. For `INSERT INTO ... SELECT`. That doesn't get converted into a `Plan::ReadThenWrite` until partway through sequencing.